### PR TITLE
 (RE-11480) Add `ship_nightly_gem` task

### PR DESCRIPTION
**Pull request by mwaggett**
In addition to some other small maintenance, this PR adds a rake task for shipping nightly gems. This task will ship the gems to Artifactory and to the nightly repos on weth. This commit also adds a `nonfinal_gem_path` param, similar to existing `nonfinal_<file type>_path` params, to distinguish between the final `gem_path` - this param will need to be added to build_defaults and/or build-data for any projects that wish to ship nightly gems.

See individual commit messages for further details.

### CODE REVIEW

**Comment by mwaggett** [comment]
Tested by building a dev version of hiera, setting NONFINAL_GEM_PATH to '/home/molly.waggett/tmp', and running the `pl:ship_nightly_gem` task from the hiera repo.

Gem was successfully shipped to Artifactory: https://artifactory.delivery.puppetlabs.net/artifactory/webapp/#/artifacts/browse/tree/General/rubygems__local/gems/hiera-3.4.5.13.gem
Gem was successfully shipped to weth: 
```
molly.waggett@weth:~/tmp$ ls
hiera-3.4.5.13.gem
```

**Comment by mwaggett** [comment]
Also tested without NONFINAL_GEM_PATH and the task failed as expected:
```
molly.waggett@molly:~/puppet-projects/hiera(master👌 ) $ bundle exec rake pl:ship_nightly_gem
Skipping platform_data collection, but don't worry about it.
touch /Users/molly.waggett/.packaging
Cloning into '/var/folders/0q/nntw0zrj1sv01qc81rtvtnbm0000gq/T/pkgXXXXXX.Nfp9hanc'...
remote: Enumerating objects: 4, done.
remote: Counting objects: 100% (4/4), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 837 (delta 0), reused 3 (delta 0), pack-reused 833
Receiving objects: 100% (837/837), 139.89 KiB | 8.74 MiB/s, done.
Resolving deltas: 100% (208/208), done.
Switched to a new branch 'release'
Warning - No build data parameter found for 'internal_nexus_host'. Perhaps you have an erroneous entry in your yaml file?
Switched to a new branch 'hiera'
rm -rf /var/folders/0q/nntw0zrj1sv01qc81rtvtnbm0000gq/T/pkgXXXXXX.Nfp9hanc
rake aborted!
Value `Pkg::Config.nonfinal_gem_path` not defined, skipping nightly ship
/Users/molly.waggett/.rbenv/versions/2.4.3/bin/bundle:23:in `load'
/Users/molly.waggett/.rbenv/versions/2.4.3/bin/bundle:23:in `<main>'
Tasks: TOP => pl:ship_nightly_gem
(See full trace by running task with --trace)
```

File: lib/packaging/gem.rb [hunk 1]
```diff
@@ -30,5 +33,11 @@ def ship_to_rubygems(file)
       puts e
       raise e
     end
+
+    def ship_to_internal_mirror(file)
+      internal_mirror_url = 'https://artifactory.delivery.puppetlabs.net/artifactory/api/gems/rubygems'
```
[the remark this answers is not in the document]
Comment (mwaggett) [reply lib/packaging/gem.rb]: ah yeah, I was thinking about that and then forgot to go look and see if it was already set somewhere.
Looks like we have the base url in https://github.com/puppetlabs/packaging/blob/1.0.x/lib/packaging/artifactory.rb - do you think we should make use of that or just add an `internal_gem_mirror` config param?

[the remark this answers is not in the document]
Comment (mwaggett) [reply lib/packaging/gem.rb]: nvm, we already have it here: https://github.com/puppetlabs/packaging/blob/1.0.x/lib/packaging/config/params.rb#L308 😊

File: lib/packaging/gem.rb [hunk 2]
```diff
@@ -30,5 +33,10 @@ def ship_to_rubygems(file)
       puts e
       raise e
     end
+
+    def ship_to_internal_mirror(file)
+      internal_mirror_api_key_name = 'artifactory_api_key'
+      ship_to_rubygems(file, { :host => Pkg::Config.internal_gem_host, :key => internal_mirror_api_key_name })
```
Comment (mwaggett) [inline lib/packaging/gem.rb]: @underscorgan updated!
re-ran my tests and they worked the same as before.

**Comment by mwaggett** [comment]
@underscorgan According to https://guides.rubygems.org/patterns/#prerelease-gems, any letters in the gem version should qualify it as a prerelease. The hiera gem I built for testing did NOT have any letters in it, just the extra digit for number of commits past the last tag. I'm gonna have to do some digging to figure out how `Pkg::Config.gemversion` gets set, but yeah, we will probably want to adjust that to include the sha? or add something like `-dev` or `-pre-release` to the version if there are four digits?

**Comment by shrug** [comment]
Where does the gem versioning live, in the project or in packaging? It's been a while since I've been in there and I don't remember

**Comment by mwaggett** [comment]
@shrug it lives in packaging: https://github.com/puppetlabs/packaging/blob/1.0.x/lib/packaging/config.rb#L231

**Comment by shrug** [comment]
I was about to say never mind, found it :) https://github.com/puppetlabs/packaging/blob/1.0.x/tasks/gem.rake

**Comment by shrug** [comment]
There's probably a bigger discussion about shipping pre-release gems to rubygems.org, but we have to sort out the versioning stuff first

**Pull request merged**